File: 56.c
#include <stdio.h>
#include <stdlib.h>

/**
 * Return an array of arrays of size *returnSize.
 * The sizes of the arrays are returned as *returnColumnSizes.
 * Note: Both ** and *returnSize and *returnColumnSizes must be malloced.
 */

int compare(const void *a, const void *b) {
    int *intervalA = *(int **)a;
    int *intervalB = *(int **)b;
    return intervalA[0] - intervalB[0];
}

int** merge(int** intervals, int intervalsSize, int* intervalsColSize, int* returnSize, int** returnColumnSizes) {
    if (intervalsSize == 0) {
        *returnSize = 0;
        *returnColumnSizes = NULL;
        return NULL;
    }
    
    // Sort intervals by start time
    qsort(intervals, intervalsSize, sizeof(int*), compare);
    
    // Allocate memory for result
    int** result = (int**)malloc(intervalsSize * sizeof(int*));
    *returnColumnSizes = (int*)malloc(intervalsSize * sizeof(int));
    
    // Add first interval
    result[0] = (int*)malloc(2 * sizeof(int));
    result[0][0] = intervals[0][0];
    result[0][1] = intervals[0][1];
    (*returnColumnSizes)[0] = 2;
    int resultSize = 1;
    
    // Merge overlapping intervals
    for (int i = 1; i < intervalsSize; i++) {
        int start = intervals[i][0];
        int end = intervals[i][1];
        int lastEnd = result[resultSize - 1][1];
        
        if (start <= lastEnd) {
            // Merge intervals
            result[resultSize - 1][1] = (end > lastEnd) ? end : lastEnd;
        } else {
            // Add new interval
            result[resultSize] = (int*)malloc(2 * sizeof(int));
            result[resultSize][0] = start;
            result[resultSize][1] = end;
            (*returnColumnSizes)[resultSize] = 2;
            resultSize++;
        }
    }
    
    *returnSize = resultSize;
    return result;
}

// Time complexity: O(n log n)
// Space complexity: O(n)

File: 56.py
'''Given an array of intervals where intervals[i] = [starti, endi], merge all overlapping intervals, and return an array of the non-overlapping intervals that cover all the intervals in the input.'''

class Solution(object):
    def merge(self, intervals):
        """
        :type intervals: List[List[int]]
        :rtype: List[List[int]]
        """
        intervals.sort(key=lambda i : i[0])
        output =[intervals[0]]
        
        for start ,end  in intervals [1:]:
            lastEnd= output [-1][1] 
            
            if start <= lastEnd:
                output [-1][1] =max (end , lastEnd)
                
            # [1,5] , [2,4] = [1,5]

            else:
                output.append([start,end])
        return output 
    